library/core/src/slice/sort/stable/quicksort.rs

```diff
@@ -0,0 +1,258 @@
+        // semantically act as moves, permuting `v`.
+
+        // Copy all the elements < p directly from swap to v.
+        let v_base = v.as_mut_ptr();
+        ptr::copy_nonoverlapping(scratch_base, v_base, state.num_left);
+
+        // Copy the elements >= p in reverse order.
+        for i in 0..len - state.num_left {
+            ptr::copy_nonoverlapping(
+                scratch_base.add(len - 1 - i),
+                v_base.add(state.num_left + i),
+                1,
+            );
+        }
+
+        state.num_left
+    }
+}
+
+struct PartitionState<T> {
+    // The start of the scratch auxiliary memory.
+    scratch_base: *mut T,
+    // The current element that is being looked at, scans left to right through slice.
+    scan: *const T,
+    // Counts the number of elements that went to the left side, also works around:
+    // https://github.com/rust-lang/rust/issues/117128
+    num_left: usize,
+    // Reverse scratch output pointer.
+    scratch_rev: *mut T,
+}
+
+impl<T> PartitionState<T> {
+    /// # Safety
+    /// scan and scratch must point to valid disjoint buffers of length len. The
+    /// scan buffer must be initialized.
+    unsafe fn new(scan: *const T, scratch: *mut T, len: usize) -> Self {
+        // SAFETY: See function safety comment.
+        unsafe { Self { scratch_base: scratch, scan, num_left: 0, scratch_rev: scratch.add(len) } }
+    }
+
+    /// Depending on the value of `towards_left` this function will write a value
+    /// to the growing left or right side of the scratch memory. This forms the
+    /// branchless core of the partition.
+    ///
+    /// # Safety
+    /// This function may be called at most `len` times. If it is called exactly
+    /// `len` times the scratch buffer then contains a copy of each element from
+    /// the scan buffer exactly once - a permutation, and num_left <= len.
+    unsafe fn partition_one(&mut self, towards_left: bool) -> *mut T {
+        // SAFETY: see individual comments.
+        unsafe {
+            // SAFETY: in-bounds because this function is called at most len times, and thus
+            // right now is incremented at most len - 1 times. Similarly, num_left < len and
+            // num_right < len, where num_right == i - num_left at the start of the ith
+            // iteration (zero-indexed).
+            self.scratch_rev = self.scratch_rev.sub(1);
+
+            // SAFETY: now we have scratch_rev == base + len - (i + 1). This means
+            // scratch_rev + num_left == base + len - 1 - num_right < base + len.
+            let dst_base = if towards_left { self.scratch_base } else { self.scratch_rev };
+            let dst = dst_base.add(self.num_left);
+            ptr::copy_nonoverlapping(self.scan, dst, 1);
+
+            self.num_left += towards_left as usize;
+            self.scan = self.scan.add(1);
+            dst
+        }
+    }
+}
+
+#[const_trait]
```

Comment:
CC @rust-lang/wg-const-eval Can we use this?

Reply:
const trait is more @rust-lang/types than wg-const-eval...  I don't know the current status

Reply:
Hi! This is @rust-lang/project-const-traits actually. I don't think our implementation is good enough to be indirectly used by any stable library function. At least not until #120639 lands.

Reply:
I'd like to add that the `const_trait` here are not a direct goal but rather a way to achieve trait impl checking in const functions. There are two places where this is used:

1. `if T::is_copy()` is used in `choose_unstable_small_sort` which is used as optimization heuristic.
2. `has_direct_interior_mutability` is used twice in `stable::quicksort` to enable an optimization. This could easily be a non-const function.

So the only thing that really needs it is `choose_unstable_small_sort`, which needs a const eval way to decide whether a type implements copy or not. It's possible to make this decision without const eval but that implies compiling all 3 possible variants, which comes with a hefty compile-time penalty I'd like to avoid.

For completeness sake, `Freeze` is used via  `#[rustc_unsafe_specialization_marker] pub(crate) trait FreezeMarker {}` to specialize the chosen implementations, because deep down the bi-directional merge in the small-sort is not sound for non `Freeze` types.

Reply:
> So the only thing that really needs it is `choose_unstable_small_sort`, which needs a const eval way to decide whether a type implements copy or not. It's possible to make this decision without const eval but that implies compiling all 3 possible variants, which comes with a hefty compile-time penalty I'd like to avoid.

You'd probably want specialization. `const_trait` should not be used.

Reply:
Specialization while sounding more appropriate doesn't solve this either, because the heuristic also takes `mem::size_of::<T>()` into consideration. It's possible to use specialization there -- [as we've done previously](https://github.com/Voultapher/sort-research-rs/commit/16e8119deb2f05407a528a1400b9de17e4bd8e2f) -- but that would require `generic_const_expr` which can't easily be turned on in the stdlib.

Specialization is the way to check if a trait is implemented, and `const_trait` is the part that makes this information available to const functions. What alternative do you suggest @fee1-dead ?

Reply:
As you said, you are also looking at `mem::size_of::<T>`, which could have perf/compile time impact considerations. My usual suggestion would be to use specialization for the logic so you can branch Freeze vs !Freeze, if that doesn't work, then it simply doesn't work. As much as I would like for `const_trait` to be experimented with uses within libcore, this intersects with specialization, and that will need more implementation work, (The WIP PR for a rewrite of that feature is already big) so this could have a negative impact on the compiler side. I can't currently think of an alternative, sorry. On the other hand, I'm looking into reviving the WIP PR and submit it for review within next week, so we might be able to have it work, if you can't think of a replacement before then.

(I did check that specializing on associated constants and types are not included within `min_specialization`)

Reply:
How about this, I remove the `const_trait` from `has_direct_interior_mutability` and leave it in `is_copy`. Then there is only one place that relies on it, and does so primarily for compile-time reasons. When refactoring the `const_trait` stuff and you run into issues because of the `is_copy` related `const_trait`, we remove it and bite the compile-time regression bullet then if we can't find an alternative at that future point in time.

The code as checked in, works as far as I can tell without holes. Not doing the `is_copy` specialization now, seems like a needless compile-time pessimization we don't need to make as of now.

Reply:
As discussed, `const_trait` is deemed not ready for this use-case, so I've replaced it with a two-stage trait specialization approach.